components/brave_vpn/brave_vpn_service.cc

```diff
@@ -0,0 +1,261 @@
+#include "base/json/json_writer.h"
+
+namespace {
+constexpr char kVpnHost[] = "housekeeping.sudosecuritygroup.com";
+
+constexpr char kAllServerRegions[] = "api/v1/servers/all-server-regions";
+constexpr char kTimezonesForRegions[] =
+    "api/v1.1/servers/timezones-for-regions";
+constexpr char kHostnameForRegion[] = "api/v1/servers/hostnames-for-region";
+constexpr char kCreateSubscriberCredential[] =
+    "api/v1/subscriber-credential/create";
+constexpr char kVerifyPurchaseToken[] = "api/v1.1/verify-purchase-token";
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("brave_vpn_service", R"(
+      semantics {
+        sender: "Brave VPN Service"
+        description:
+          "This service is used to communicate with Guardian VPN apis"
+          "on behalf of the user interacting with the Brave VPN."
+        trigger:
+          "Triggered by user connecting the Brave VPN."
+        data:
+          "Servers, hosts and credentials for Brave VPN"
+        destination: WEBSITE
+      }
+    )");
+}
+
+GURL GetURLWithPath(const std::string& host, const std::string& path) {
+  return GURL(std::string(url::kHttpsScheme) + "://" + host).Resolve(path);
+}
+
+std::string CreateJSONRequestBody(const base::Value& dict) {
+  std::string json;
+  base::JSONWriter::Write(dict, &json);
+  return json;
+}
+
+std::string GetSubscriberCredentialFromJson(const std::string& json) {
+  base::JSONReader::ValueWithError value_with_error =
+      base::JSONReader::ReadAndReturnValueWithError(
+          json, base::JSONParserOptions::JSON_PARSE_RFC);
+  base::Optional<base::Value>& records_v = value_with_error.value;
+  if (!records_v) {
+    LOG(ERROR) << "Invalid response, could not parse JSON, JSON is: " << json;
+    return "";
+  }
+
+  const base::DictionaryValue* response_dict;
+  if (!records_v->GetAsDictionary(&response_dict)) {
+    return "";
+  }
+  
+  const base::Value* subscriber_credential =
+      records_v->FindKey("subscriber-credential");
+  return subscriber_credential == nullptr ? ""
+                                          : subscriber_credential->GetString();
+}
+
+}  // namespace
+
+BraveVpnService::BraveVpnService(scoped_refptr<network::SharedURLLoaderFactory> url_loader_factory)
+    : url_loader_factory_(url_loader_factory),
+      weak_factory_(this) {}
+
+BraveVpnService::~BraveVpnService() = default;
+
+void BraveVpnService::OAuthRequest(const GURL& url,
+                                   const std::string& method,
+                                   const std::string& post_data,
+                                   bool set_app_ident,
+                                   URLRequestCallback callback) {
+  auto request = std::make_unique<network::ResourceRequest>();
+  request->url = url;
+  request->method = method;
+  if (set_app_ident) {
+    request->headers.SetHeader("GRD-App-Ident", "Brave-Client");
+  }
+  auto url_loader = network::SimpleURLLoader::Create(
```

Comment:
there is a new utility called `BraveApiRequest` that probably could be used here to reduce boilerplate